File: rust/herb-linter/src/rules/erb_no_then_in_control_flow.rs
use herb::nodes::{ERBIfNode, ERBInNode, ERBUnlessNode, ERBWhenNode};
use herb::Location;
use herb::Visitor;

rule_visitor!(ERBNoThenInControlFlowVisitor);
define_parser_rule!(
  ERBNoThenInControlFlowRule,
  "erb-no-then-in-control-flow",
  Warning,
  ERBNoThenInControlFlowVisitor,
  parser_options: { strict: true }
);

impl ERBNoThenInControlFlowVisitor {
  fn check_then_keyword(&mut self, keyword: &str, then_keyword: &Option<Location>) {
    if let Some(location) = then_keyword {
      self.add_offense(
        format!(
          "Avoid using `then` in `{}` expressions inside ERB templates. Use the multiline block form instead.",
          keyword
        ),
        location.clone(),
      );
    }
  }
}

impl Visitor for ERBNoThenInControlFlowVisitor {
  fn visit_erb_if_node(&mut self, node: &ERBIfNode) {
    let content = node.content.as_ref().map(|token| token.value.trim()).unwrap_or("");
    let keyword = if content.starts_with("elsif") { "elsif" } else { "if" };

    self.check_then_keyword(keyword, &node.then_keyword);
    self.walk_erb_if_node(node);
  }

  fn visit_erb_unless_node(&mut self, node: &ERBUnlessNode) {
    self.check_then_keyword("unless", &node.then_keyword);
    self.walk_erb_unless_node(node);
  }

  fn visit_erb_when_node(&mut self, node: &ERBWhenNode) {
    self.check_then_keyword("when", &node.then_keyword);
    self.walk_erb_when_node(node);
  }

  fn visit_erb_in_node(&mut self, node: &ERBInNode) {
    self.check_then_keyword("in", &node.then_keyword);
    self.walk_erb_in_node(node);
  }
}

File: javascript/packages/linter/src/rules/erb-no-then-in-control-flow.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ERBIfNode, ERBUnlessNode, ERBWhenNode, ERBInNode, ParseResult, ParserOptions, Location } from "@herb-tools/core"

class ERBNoThenInControlFlowVisitor extends BaseRuleVisitor {
  visitERBIfNode(node: ERBIfNode): void {
    const content = node.content?.value?.trim() ?? ""
    const keyword = content.startsWith("elsif") ? "elsif" : "if"

    this.checkThenKeyword(keyword, node.then_keyword)
    this.visitChildNodes(node)
  }

  visitERBUnlessNode(node: ERBUnlessNode): void {
    this.checkThenKeyword("unless", node.then_keyword)
    this.visitChildNodes(node)
  }

  visitERBWhenNode(node: ERBWhenNode): void {
    this.checkThenKeyword("when", node.then_keyword)
    this.visitChildNodes(node)
  }

  visitERBInNode(node: ERBInNode): void {
    this.checkThenKeyword("in", node.then_keyword)
    this.visitChildNodes(node)
  }

  private checkThenKeyword(keyword: string, thenKeyword: Location | null): void {
    if (thenKeyword === null) return

    this.addOffense(
      `Avoid using \`then\` in \`${keyword}\` expressions inside ERB templates. Use the multiline block form instead.`,
      thenKeyword,
    )
  }
}

export class ERBNoThenInControlFlowRule extends ParserRule {
  static ruleName = "erb-no-then-in-control-flow"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning",
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return { strict: true }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoThenInControlFlowVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
